### ISSUE TITLE
Monthly Milestone - February 2019 💝⛄️

### ISSUE THREAD

**Issue by dhmlau** [strongloop/loopback-next#2313]
## February milestone
- Authentication & Authorization story
- Spikes on LB3->LB4 Migration
- Groundwork for LB4 extensibility
- Prep work for presentations at events

Note: 
   - _We have committed more than our usual velocity because some are almost done, and some need to happen (e.g. conference prep work)._
   - _We normally don't include PRs in the milestone, but the extensibility PRs are laying the foundation and already taking a lot of the team's time for review._

### LoopBack 4
- [ ] Authentication & Authorization #1035
  - [x] Store hashed password in the login function https://github.com/strongloop/loopback4-example-shopping/issues/35
  - [x] Refactor authentication util functions into a service strongloop/loopback4-example-shopping#40

- [ ] Improve LB4 extensibility PRs
  - [ ] 🚶feat(context): add support to watch context bindings, #2122
  - [ ] 🚶feat(context): support binding config and @inject.config, #2259
  - [ ] 🚶[RFC] Greeter extension to illustrate extension point/extension pattern, #2249
  - [ ] 🚶feat(core): introduce basic life cycle support for LoopBack applications, #1928
  - [ ] 🚶feat(events): add basic async events support, #1972

- [ ] Migration https://github.com/strongloop/loopback-next/issues/1849
  - [x] Continue the work on the original spike #2224
  - [x] Spike: Mount LB3 app and include its REST API in OpenAPI spec #2301

- [x] User adoption
   - [x] Spike: How to write a LB3-like application in JavaScript, https://github.com/strongloop/loopback-next/issues/1978.  Last iteration of the spike.
   - [x] Tutorial/Blog to show how to mount LB app to Express/Koa as middleware, #1982

- [x] Deployment to IBM Cloud
   - [x] [Spike] k8s based deployment with IBM Cloud #1606

- [x] Conference preparation
   - [x] Update Quicklab for Code@Think  https://github.com/strongloop/loopback-next/issues/2253
   - [x] Prep work / presentation at DeveloperWeek 2019  & THINK 2019

- [ ] Bugs
  - [ ] [Spike] Partial update (PATCH) over REST #2082

- [ ] Others
   - [ ] APIConnect 5.0.8.6 toolkit and collective testing https://github.com/strongloop-internal/scrum-apex/issues/405
   - [x] 🚶Step-by-step instructions to submit a PR, https://github.com/strongloop/loopback-next/issues/2280

### Stretch Goals
- [ ] Authentication
   - [ ] ~Extension point for register passport strategies  https://github.com/strongloop/loopback-next/issues/2311~ Moved to Q2
   - [ ] Extension point for register non-passport actions https://github.com/strongloop/loopback-next/issues/2312
   - [ ] Make `UserProfile` more flexible https://github.com/strongloop/loopback-next/issues/2246
   - [ ] Add logout endpoint in the user controller https://github.com/strongloop/loopback4-example-shopping/issues/36

- [ ] Bug:
  - [ ] Mongo fails to find user by id (hasMany relation not working) #2085
  - [x] [Spike] hasMany relation to same Entity fails #1682
  - [ ] [Spike] Explorer the way to resolve inclusion property https://github.com/strongloop/loopback-next/issues/2152. 
  - [ ] Partial update (PATCH) over REST, https://github.com/strongloop/loopback-next/issues/1722. 
  - [ ] hasMany relation to same Entity fails #1571

### LoopBack 3
- [ ]  🚶Fix the vulnerability issue caused by swagger-ui@2.x https://github.com/strongloop/loopback-component-explorer/issues/254

**Comment by jannyHou** [comment]
@dhmlau @nabdelgadir According to the estimation meeting yesterday, https://github.com/strongloop/loopback-next/issues/2311 is blocked by https://github.com/strongloop/loopback-next/pull/2249, I propose we work on https://github.com/strongloop/loopback4-example-shopping/issues/40 as a replacement, there is already a PR in progress for it: https://github.com/strongloop/loopback4-example-shopping/pull/33
what would you think?

**Comment by nabdelgadir** [comment]
@jannyHou sounds good to me. @dhmlau what do you think?

**Comment by dhmlau** [comment]
@nabdelgadir, @jannyHou and I were just discussing this, and we've modified the milestone.

**Comment by dhmlau** [comment]
Based on the discussion summarized in [here](https://github.com/strongloop/loopback-next/issues/1035#issuecomment-465311972), the next step for the authentication story is 

>* **Step 1a**: create extension point for access token generator in `@loopback/authentication`, define the key as the constant and interface
  
>   * in shopping example, it will implement this access token generator
>* **Step 1b**: create interface for strategy and then the extension point for strategies -- needs Raymond's PR
>* **Step 1c**: Add common action to authenticate inside authentication module.
>* ** Step 1d**: Move the credential extractor code from the example app back to the `authentication` package.
>* **Step 2** : rebase shopping example against the `@loopback/authentication`

So the stretch goals under authentication epic won't be carried over to March, and @jannyHou will be creating new set of tasks for the above.
